### PR TITLE
8332042: Move MEMFLAGS to its own include file

### PR DESCRIPTION
**Pull request by tstuefe**
MEMFLAGS, as well as its enum constants, should live in its own include. 

The constants are used throughout the code base, often without needing the allocation APIs exposed through allocation.hpp.

The MEMFLAGS enum def is often needed within NMT itself, again often without needing allocation.hpp.

---

This patch moves the enum to its new file.

It fixes those `allocation.hpp` includes that where only needed to get MEMFLAGS. It does not fix other includes. 

For backward compatibility, until we straightened out the dependencies (e.g., fixing all places where we rely on indirect includes), I added memflags.hpp to allocation.hpp.

I tested (built) on:
- MacOS aarch64, no precompiled headers, fastdebug
- Linux x64, no precompiled headers, fastdebug, release, fastdebug crossbuild to aarch64, fastdebug minimal

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332042](https://bugs.openjdk.org/browse/JDK-8332042): Move MEMFLAGS to its own include file (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [42361558](https://git.openjdk.org/jdk/pull/19172/files/42361558b28e154e6bb9e22e82405c35f6b7d348)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19172/head:pull/19172` \
`$ git checkout pull/19172`

Update a local copy of the PR: \
`$ git checkout pull/19172` \
`$ git pull https://git.openjdk.org/jdk.git pull/19172/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19172`

View PR using the GUI difftool: \
`$ git pr show -t 19172`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19172.diff">https://git.openjdk.org/jdk/pull/19172.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19172#issuecomment-2104353539)

### CODE REVIEW

**Comment by tstuefe** [comment]
Ping @afshin-zafari @jdksjolen 
@gerard-ziemski

File: src/hotspot/share/services/mallocLimit.hpp
Comment (afshin-zafari) [inline src/hotspot/share/services/mallocLimit.hpp:3]: Maybe 2023, 2024 instead?

**Review by jdksjolen — APPROVED** [review]
Seems reasonable and LGTM, thanks.

File: src/hotspot/share/nmt/mallocTracker.hpp
Comment (stefank) [inline src/hotspot/share/nmt/mallocTracker.hpp:29]: Should go after mallocHeader.hpp

File: src/hotspot/share/nmt/memflags.cpp
Comment (stefank) [inline src/hotspot/share/nmt/memflags.cpp:27]: There should be no blankline between precompiled.hpp and the rest of the includes.

File: src/hotspot/share/nmt/memflags.cpp
Comment (stefank) [inline src/hotspot/share/nmt/memflags.cpp:31]: I think you can remove this entire .cpp file. There's no need to check the size of an enum with a specified base type.

File: src/hotspot/share/nmt/memflags.hpp
Comment (stefank) [inline src/hotspot/share/nmt/memflags.hpp:30]: Open-ended comment/question: We call it MEMORY_TYPE and mt, but then we call the type MEMFLAGS (with a completely non-standard UPPERCASE style). Maybe it is time to rename MEMFLAGS?

File: src/hotspot/share/services/mallocLimit.cpp
Comment (stefank) [inline src/hotspot/share/services/mallocLimit.cpp:28]: While poking around in the includes, could you remove the blankline on 27. This style inconsistency has slowly crept into the code base.

**Review by stefank — CHANGES_REQUESTED** [review]

File: src/hotspot/share/nmt/memflags.cpp
Comment (tstuefe) [reply src/hotspot/share/nmt/memflags.cpp:31]: I rather have this explicit check. If MEMFLAGS>1byte, things break, and I would like to make that explicit.

That said, I can move this static assert to the header. I just wanted to avoid including debug.hpp. Note that the cpp file would be the right place in the future for any MEMFLAGS related utility functions, e.g. to-and-from-string conversations.

Comment (stefank) [reply src/hotspot/share/nmt/memflags.cpp:31]: Could you instead put the static_assert near the code that will break? Right now it looks obscure and weird to have this check when it is obviously correct as long as no one changes the definition. Would it be enough to write a comment in the header that this needs to be 1 byte?

File: src/hotspot/share/nmt/memflags.hpp
Comment (stefank) [reply src/hotspot/share/nmt/memflags.hpp:30]: Right. That's why I prefixed this with "Open-ended comment/question", trying to make it super clear that it wasn't intended as a request for this PR, but rather a way to at least plant the seed of an idea that we might want to fix this eyesore.

File: src/hotspot/share/nmt/memflags.cpp
Comment (dcubed-ojdk) [reply src/hotspot/share/nmt/memflags.cpp:31]: To quote @robehn - Why write a comment for a rule if you can enforce it with code instead...

Comment (stefank) [reply src/hotspot/share/nmt/memflags.cpp:31]: I tend to agree with that. My earlier question still stands is there a better place to put it? Right now the "enforced" code in a stand-alone file doesn't tell me "why" this is important.

Comment (stefank) [reply src/hotspot/share/nmt/memflags.cpp:31]: If you want to keep the static_assert it in the .cpp file, then I won't block that.

**Review by kimbarrett — COMMENTED** [review]
Looks good, subject to addressing the minor issues already reported by others.
Leaving unapproved, but don't wait for my approval once you have others.

File: src/hotspot/share/nmt/memflags.cpp
Comment (tstuefe) [reply src/hotspot/share/nmt/memflags.cpp:31]: > Could you instead put the static_assert near the code that will break?
I like that. I will do that.

File: src/hotspot/share/nmt/memflags.hpp
Comment (tstuefe) [reply src/hotspot/share/nmt/memflags.hpp:30]: I agree with you on the eyesore. 

MEMFLAGS does not follow any established convention, the implied plural is strange (its just one flag, not a set of), etc. We will change it sometime in the future.

File: src/hotspot/share/nmt/memflags.cpp
Comment (tstuefe) [reply src/hotspot/share/nmt/memflags.cpp]: I removed the file.

**Comment by tstuefe** [comment]
@stefank New version, hopefully addressed all your remarks. Thanks!

**Review by stefank — APPROVED** [review]
Looks good. Thanks!

**Comment by tstuefe** [comment]
Thanks @afshin-zafari @stefank @kimbarrett @jdksjolen 

/integrate

**Pull request closed without merging**
